File: Session 5 - Chain Table/Uploaded files/Java/DoublyLinkedList.java
class Node{
    int key;
    Node prev;
    Node next;
    Node(int value){
        prev = null;
        next = null;
        key = value;
    }
}

class DoublyLinkedList {
    Node head = null;
    Node tail = null;

    void insertBefore(Node y, Node x){
       // TODO
    }

    void insertTail(Node x){
        if (this.tail == null){
            this.head = x;
            this.tail = x;
        }else {
            this.tail.next = x;
            x.prev = this.tail;
            this.tail = x;
        }
    }

    void insertKeepSorted(Node x){
        Node p = this.head;
        while (p != null && p.key < x.key){
            p = p.next;
        }
        if (p == null){
            insertTail(x);
        }else {
            insertBefore(p, x);
        }
    }


    Node search(int value){
        Node node = head;
        while (node != null){
            if (node.key == value){
                break;
            }
            node = node.next;
        }
        return node;
    }

    void delete(Node node) {
        if (head == null || node == null) {
            return;
        }

        if (head == node) {
            head = node.next;
        }

        if (tail == node){
            tail = node.prev;

        }

        if (node.next != null) {
            node.next.prev = node.prev;
        }

        if (node.prev != null) {
            node.prev.next = node.next;
        }

        return;
    }
}

public class Main{
    public static void main(String[] args) {
       // TODO
    }
}

File: Session 5 - Chain Table/Uploaded files/Python/DoublyLinkedList.py
class Node:
    def __init__(self, value):
        self.key = value
        self.prev = None
        self.next = None

class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None

    def insert_before(self, y, x):
        pass
        # TODO

    def insert_tail(self, x):
        if self.tail is None:
            self.head = x
            self.tail = x
        else:
            self.tail.next = x
            x.prev = self.tail
            self.tail = x

    def insert_keep_sorted(self, x):
        p = self.head
        while p is not None and p.key < x.key:
            p = p.next

        if p is None:
            self.insert_tail(x)
        else:
            self.insert_before(p, x)

    def search(self, value):
        node = self.head
        while node is not None:
            if node.key == value:
                return node
            node = node.next
        return None

    def delete(self, node):
        if self.head is None or node is None:
            return

        if self.head == node:
            self.head = node.next

        if self.tail == node:
            self.tail = node.prev

        if node.next is not None:
            node.next.prev = node.prev

        if node.prev is not None:
            node.prev.next = node.next

# Example usage
if __name__ == "__main__":
    pass
    # TODO
